src/expr/src/relation/mod.rs

```diff
@@ -455,6 +455,162 @@ impl RelationExpr {
         self.typ().column_types.len()
     }
 
+    /// Recursively descend two expressions until differences are found.
+    ///
+    /// The resulting sequence contains the first moments at which differences were observed.
+    /// There may be multiple elements in the list as the differences may only be observed
+    /// when descending independent tree paths.
+    pub fn tree_differences(&self, other: &RelationExpr) -> Vec<(RelationExpr, RelationExpr)> {
```

Comment:
I deleted it because it was too annoying to continually rebase, but it is a much clearer tree difference between expressions. When the expression is tens+ of nodes, but only a few changed, it just reveals those changes. But the comments say all of that. It was useful in debugging transforms.